pkg/controller/registry/grpc/source.go

```diff
@@ -0,0 +1,217 @@
+package grpc
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/connectivity"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/registry/resolver"
+	"github.com/operator-framework/operator-registry/pkg/api"
+	"github.com/operator-framework/operator-registry/pkg/api/grpc_health_v1"
+	"github.com/operator-framework/operator-registry/pkg/client"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type SourceMeta struct {
+	Address         string
+	LastConnect     metav1.Time
+	ConnectionState connectivity.State
+}
+
+type SourceState struct {
+	Key   resolver.CatalogKey
+	State connectivity.State
+}
+
+type SourceConn struct {
+	SourceMeta
+	Conn   *grpc.ClientConn
+	cancel context.CancelFunc
+}
+
+type SourceStore struct {
+	sync.Once
+	sources      map[resolver.CatalogKey]SourceConn
+	sourcesLock  sync.RWMutex
+	syncFn       func(SourceState)
+	logger       *logrus.Logger
+	notify       chan SourceState
+	timeout      time.Duration
+	readyTimeout time.Duration
+}
+
+func NewSourceStore(logger *logrus.Logger, timeout, readyTimeout time.Duration, sync func(SourceState)) *SourceStore {
+	return &SourceStore{
+		sources:      make(map[resolver.CatalogKey]SourceConn),
+		notify:       make(chan SourceState),
+		syncFn:       sync,
+		logger:       logger,
+		timeout:      timeout,
+		readyTimeout: readyTimeout,
+	}
+}
+
+func (s *SourceStore) Start(ctx context.Context) {
```

Comment:
nit: Extra "empty" goroutines can be created. Not sure if the compiler will optimize them away, but you could always create the goroutine _inside_ the `Do` function and avoid this all together:
  
```suggestion
func (s *SourceStore) Start(ctx context.Context) {
    s.logger.Debug("starting source manager")
    s.Do(func() {
        // Only start one source manager goroutine in the lifetime of s
	go func() {
	    for {
                select {
                case <-ctx.Done():
	           s.logger.Debug("closing source manager")
	           return
	        case e := <-s.notify:
                    s.logger.Debugf("Got source event: %#v", e)
	            s.syncFn(e)
	        }
	    }
	}()
    }
}
```

Reply:
The “extra” goroutines would immediately exit

Reply:
True - not a big deal at all - but why bother creating them in the first place?